contrib/startup/GafferUI/layoutCompatibility.py

```diff
@@ -40,23 +40,45 @@
 import GafferUI
 
 import imath
+import inspect
 
-if Gaffer.About.compatibilityVersion() < 54 :
+# The signature for CompoundEditor.__init__ has changed some:
+#  - 0.54.0.0 introduced 'windowState' and 'detachedPanels'
+#  - 0.x.x.x migrated to the opaque _state dict
+# We need to patch the constructor such that it only receives the args that
+# it expects
+
+initSpec = inspect.getargspec( GafferUI.CompoundEditor.__init__ )
+
+def __initWrapper( originalInit ) :
+
+	def init( self, *args, **kwargs ) :
+
+		# If this version uses _state and incoming args are pre-state, nest them
+		if "_state" in initSpec.args and "_state" not in kwargs :
+			kwargs[ "_state" ] = kwargs
+		# Or, if it doesn't and we do have them, we need to unpack them
+		elif "_state" not in initSpec.args and "_state" in kwargs :
+			kwargs.update( kwargs["_state"] )
 
-	# Remove new kwargs introduced to store positions/detached panels
-	def __initWrapper( originalInit ) :
+		# As we only understand specific kwargs, delete any we don't know about
+		# We never used **kwarg anywhere to pass args down so this should be safe
+		kwargs = { k:v for k, v in kwargs.items() if k in initSpec.args }
```

Comment:
I think this *might* make compatibility horrendous if we have to support `**kw`. As we need this compatibility shim to prevent local _and_ parent classes from barfing.  By not supporting it, we can simply introspect the locally supported args, we then don't need the version/arg matrix got all possible versions to know what should be there and the shim works across everything. There are already 3 permutations from 53 to this. Given we said we wanted to consider 'anything other than a default construct (with scriptNode) of the CompondEditor to be considered 'private', how important is honouring the widget convention for `**kw`?

Reply:
Im talking rubbish - we can still introspect things with `**kw`, so all is well.

Reply:
https://github.com/GafferHQ/gaffer/pull/3360/commits/f5dfb0f194c75f5732739bbc97e2339d566f4877 reinstates this.